### PR TITLE
Add validation support for text-input

### PR DESCRIPTION
**Pull request by nicknisi**
**Type:** feature

The following has been addressed in the PR:

* [x] There is a related issue
* [x] All code matches the [style guide](https://github.com/dojo/meta/blob/master/STYLE.md)
* [x] Unit or Functional tests are included in the PR

**Description:**

This adds validation support as partially described in #616. The user can provide a `validate` property to the widget, which if `true`, will validate against the `type` (`"email"`, for example) and use the browser's built-in validation. 

Alternatively, `validate` can be a function that accepts the widget's current value as an argument and must return `{ valid: boolean; message: string | undefined; }`. The user can also provide an `onValidate(valid: boolean, message?: string)` which will be called to pass the validation value back up to the parent, along with any potential message.

With the current behavior, if `validate` is a function, it is the primary validation, but the other validation (email, for example) is also being checked.

**Example:**

```typescript
w(TextInput, {
	key: 't2',
	type: 'email',
	label: 'Email (required)',
	required: true,
	// validate: true,
	validate(value: string) {
		const valid = !!~value.indexOf('nicknisi');
		return {
			valid,
			message: !valid ? 'Must include nicknisi' : ''
		};
	},
	onValidate(valid: boolean, message: string | undefined) {
		const color = valid ? 'green' : 'red';
		console.log(`%c${message}`, `font-weight:bold;color:${color};`, { valid });
	},
	value: this._value2,
	onInput: (value: string) => {
		this._value2 = value;
		this.invalidate();
	}
})
```

### Breaking changes

This PR removes the `invalid` property. The way to mark an input as invalid is to use custom validation. This also slightly affects combobox and enhanced text input as their `invalid` properties will no longer be passed down to the underlying text-input widget.

### CODE REVIEW

File: src/text-input/index.ts [hunk 1]
```diff
@@ -11,9 +11,12 @@ import { uuid } from '@dojo/framework/core/util';
 import * as css from '../theme/text-input.m.css';
 import { customElement } from '@dojo/framework/widget-core/decorators/customElement';
 import diffProperty from '@dojo/framework/widget-core/decorators/diffProperty';
+import ValidityMeta from './ValidityMeta';
 
 export type TextInputType = 'text' | 'email' | 'number' | 'password' | 'search' | 'tel' | 'url';
 
+type TextInputInternalState = any;
```
Comment (tomdye) [inline src/text-input/index.ts]: does this need typing here like this?

File: src/text-input/index.ts [hunk 2]
```diff
@@ -166,13 +202,14 @@ export class TextInputBase<P extends TextInputProperties = TextInputProperties>
...
 		} = this.properties;
+		const { valid } = this._state;
 		const focus = this.meta(Focus).get('root');
 		return [
 			css.root,
 			disabled ? css.disabled : null,
 			focus.containsFocus ? css.focused : null,
-			invalid === true ? css.invalid : null,
-			invalid === false ? css.valid : null,
+			invalid === true || valid === false ? css.invalid : null,
```
Comment (tomdye) [inline src/text-input/index.ts]: If we're adding this we will want to remove support for passing `invalid` manually.

File: src/text-input/ValidityMeta.ts [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+		}
+
+		return {
+			valid: node.validity.valid,
+			message: node.validationMessage
+		};
+	}
+}
+
+export default InputValidity;
```
Comment (tomdye) [inline src/text-input/ValidityMeta.ts]: perhaps move this to `src/common` for now?

File: src/text-input/index.ts [hunk 4]
```diff
@@ -27,16 +37,19 @@ export type TextInputType = 'text' | 'email' | 'number' | 'password' | 'search'
  * @property value           The current value
  */
 
-export interface TextInputProperties extends ThemedProperties, InputProperties, FocusProperties, LabeledProperties, PointerEventProperties, KeyEventProperties, InputEventProperties, CustomAriaProperties {
+export interface TextInputProperties extends ThemedProperties, Omit<InputProperties, 'invalid'>, FocusProperties, LabeledProperties, PointerEventProperties, KeyEventProperties, InputEventProperties, CustomAriaProperties {
```
Comment (tomdye) [inline src/text-input/index.ts]: May well be better if you just inline the appropriate properties from `InputProperties` here. Probably going to phase out all these extended properties

File: src/text-input/index.ts [hunk 5]
```diff
@@ -152,27 +165,54 @@ export class TextInputBase<P extends TextInputProperties = TextInputProperties>
...
+			return;
+		}
+
+		let { valid, message } = this.meta(ValidityMeta).get('input', value);
 
-	constructor() {
-		super();
-		this._uuid = uuid();
+		if (typeof validate === 'function') {
+			const { valid: customValid, message: customMessage } = validate(value);
```
Comment (tomdye) [inline src/text-input/index.ts]: do these need typing here given that `validate` is typed?

File: src/text-input/index.ts [hunk 6]
```diff
@@ -231,34 +272,60 @@ export class TextInputBase<P extends TextInputProperties = TextInputProperties>
 		});
 	}
 
+	protected renderHelperText(): DNode {
+		const { properties: { helperText }, _state: { message } } = this;
+		const text = message || helperText;
+		const invalid = !!message;
```
Comment (tomdye) [inline src/text-input/index.ts]: why don't you use the `_state.valid` boolean here?

File: src/text-input/index.ts (hunk 5 above)
Comment (nicknisi) [reply src/text-input/index.ts]: Well, `validate` could be a boolean, a function, or `undefined`, so I need to type guard it and treat it and call it if it is a function.

File: src/text-input/ValidityMeta.ts (hunk 3 above)
Comment (nicknisi) [reply src/text-input/ValidityMeta.ts]: Moved!

File: src/text-input/index.ts (hunk 4 above)
Comment (nicknisi) [reply src/text-input/index.ts]: Done. I copied `InputProperties` into this interface and then removed the `Omit`.

File: src/text-input/index.ts (hunk 6 above)
Comment (nicknisi) [reply src/text-input/index.ts]: Fixed.

File: src/text-input/README.md [hunk 7]
```diff
@@ -20,8 +20,10 @@ If the `label` property is not used, we recommend creating a separate `label` an
 ```typescript
 // Basic usage
 w(TextInput, {
```
Comment (tomdye) [inline src/text-input/README.md]: can you search for other mentions of `invalid` in this file and correct them please?

File: src/text-input/index.ts [hunk 8]
```diff
@@ -231,34 +273,59 @@ export class TextInputBase<P extends TextInputProperties = TextInputProperties>
 		});
 	}
 
+	protected renderHelperText(): DNode {
+		const { properties: { helperText }, _state: { valid, message } } = this;
+		const text = message || helperText;
+
+		return text ? v('div', {
```
Comment (tomdye) [inline src/text-input/index.ts]: should this perhaps be an `aside` and the text itself a `span` ? could be wrong, but just thinking out loud.

File: src/text-input/tests/unit/TextInput.ts [hunk 9]
```diff
@@ -7,21 +7,32 @@ import { v, w } from '@dojo/framework/widget-core/d';
...
 import Label from '../../../label/index';
-import TextInput from '../../index';
+import TextInput, { TextInputProperties } from '../../index';
+import InputValidity from '../../../common/InputValidity';
 import * as css from '../../../theme/text-input.m.css';
 import { compareForId, compareId, createHarness, MockMetaMixin, noop, stubEvent } from '../../../common/tests/support/test-helpers';
 
 const harness = createHarness([ compareId, compareForId ]);
 
+const sleep = (time: number = 0) => new Promise((resolve) => setTimeout(resolve, time));
```
Comment (tomdye) [inline src/text-input/tests/unit/TextInput.ts]: I don't believe this is used anywhere?

Comment (nicknisi) [reply src/text-input/tests/unit/TextInput.ts]: Deleted.

File: src/text-input/index.ts (hunk 8 above)
Comment (nicknisi) [reply src/text-input/index.ts]: Either is fine with me. I was just using the same markup as in the example [here](https://material.io/develop/web/components/input-controls/text-field/).
screen shot 2019-02-25 at 1 51 46 pm

File: src/text-input/README.md (hunk 7 above)
Comment (nicknisi) [reply src/text-input/README.md]: I also added a small section on validation to try and make things a bit more clear.

**Comment by nicknisi** [comment]
@tomdye I made changes to how Combobox internally uses `invalid`. Instead of just passing its `invalid` value down to the underlying TextInput, it instead sets `validate` to a function that passes the inverse of `invalid` if `invalid` is a boolean. It does nothing if `invalid` is `undefined`.

File: src/text-input/index.ts (hunk 8 above)
Comment (tomdye) [reply src/text-input/index.ts]: fair enough then!

**Pull request merged**